castai/resource_workload_scaling_policy.go

```diff
@@ -0,0 +1,342 @@
+				ValidateDiagFunc: validation.ToDiagFunc(validation.FloatBetween(0.01, 1)),
+			},
+		},
+	}
+}
+
+func resourceWorkloadScalingPolicyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+
+	clusterID := d.Get(FieldClusterID).(string)
+	req := sdk.WorkloadOptimizationAPICreateWorkloadScalingPolicyJSONRequestBody{
+		Name:      d.Get("name").(string),
+		ApplyType: sdk.WorkloadoptimizationV1ApplyType(d.Get("apply_type").(string)),
+		RecommendationPolicies: sdk.WorkloadoptimizationV1RecommendationPolicies{
+			ManagementOption: sdk.WorkloadoptimizationV1ManagementOption(d.Get("management_option").(string)),
+		},
+	}
+
+	if v, ok := d.GetOk("cpu"); ok {
+		req.RecommendationPolicies.Cpu = toResourcePolicies(v.([]interface{})[0].(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("memory"); ok {
+		req.RecommendationPolicies.Memory = toResourcePolicies(v.([]interface{})[0].(map[string]interface{}))
+	}
+
+	resp, err := client.WorkloadOptimizationAPICreateWorkloadScalingPolicyWithResponse(ctx, clusterID, req)
+	if checkErr := sdk.CheckOKResponse(resp, err); checkErr != nil {
+		return diag.FromErr(checkErr)
+	}
+
+	d.SetId(resp.JSON200.Id)
+
+	return resourceWorkloadScalingPolicyRead(ctx, d, meta)
+}
+
+func resourceWorkloadScalingPolicyRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+
+	clusterID := d.Get(FieldClusterID).(string)
+	resp, err := client.WorkloadOptimizationAPIGetWorkloadScalingPolicyWithResponse(ctx, clusterID, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if !d.IsNewResource() && resp.StatusCode() == http.StatusNotFound {
+		log.Printf("[WARN] Scaling policy (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+	if err := sdk.CheckOKResponse(resp, err); err != nil {
+		return diag.FromErr(err)
+	}
+
+	sp := resp.JSON200
+
+	if err := d.Set("name", sp.Name); err != nil {
+		return diag.FromErr(fmt.Errorf("setting name: %w", err))
+	}
+	if err := d.Set("apply_type", sp.ApplyType); err != nil {
+		return diag.FromErr(fmt.Errorf("setting apply type: %w", err))
+	}
+	if err := d.Set("management_option", sp.RecommendationPolicies.ManagementOption); err != nil {
+		return diag.FromErr(fmt.Errorf("setting management option: %w", err))
+	}
+	if err := d.Set("cpu", toResourceMap(sp.RecommendationPolicies.Cpu)); err != nil {
+		return diag.FromErr(fmt.Errorf("setting cpu: %w", err))
+	}
+	if err := d.Set("memory", toResourceMap(sp.RecommendationPolicies.Memory)); err != nil {
+		return diag.FromErr(fmt.Errorf("setting memory: %w", err))
+	}
+
+	return nil
+}
+
+func resourceWorkloadScalingPolicyUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if !d.HasChanges(
+		"name",
+		"apply_type",
+		"management_option",
+		"cpu",
+		"memory",
+	) {
+		log.Printf("[INFO] scaling policy up to date")
```

Comment:
Not sure why but in few other places we use `github.com/hashicorp/terraform-plugin-log/tflog` instead. I leave it up to you to evaluate if it makes sense to use it here.